src/op_analytics/cli/subcommands/pulls/agora/delegates.py

```
@@ -0,0 +1,232 @@
+                log.warning(f"Failed to fetch data for offsets {lo} or {hi}.")
+                break
+
+            print(f"max offset search range {lo} -- {hi}")
+
+            diff = hi - lo
+            midpoint = lo + (diff) // 2
+
+            if resp_lo.has_next and resp_hi.has_next:
+                lower_bound.add(hi)
+                lo = max(lower_bound)
+                hi = min(upper_bound) if upper_bound else hi * 10
+            elif resp_lo.has_next and not resp_hi.has_next:
+                lower_bound.add(lo)
+                upper_bound.add(hi)
+                if diff < 1000:
+                    self.min_offset = lo
+                    self.max_offset = hi
+                    return hi
+                else:
+                    hi = midpoint
+            else:
+                log.warning("Unexpected state encountered in binary search.")
+                hi = midpoint
+
+    def concurrent_fetch(self):
+        self.max_offset = self.binary_search_max_offset()
+        offsets = list(range(0, self.max_offset, self.limit))[:10]
+        data = run_concurrently(
+            function=lambda offset: self.request(offset).data,
+            targets=offsets,
+            max_workers=self.max_workers,
+        )
+        return list(itertools.chain.from_iterable(data.values()))
+
+
+def pull_delegates():
+    """Pull and write agora delegates data."""
+    p = Paginator(url=f"{BASE_URL}/delegates", params={"sort": "voting_power"}, max_workers=32)
+    delegate_data = p.concurrent_fetch()
+
+    # Clean and transform to Pandas
+    normalized_data = []
+    for item in delegate_data:
+        # Extract fields from the item dictionary
+        address = item.get("address", "")
+        voting_power = item.get("votingPower", {})
+        citizen = item.get("citizen", False)
+        statement = item.get("statement", {})
+
+        # Make sure voting_power is a dictionary
+        if not isinstance(voting_power, dict):
+            continue
```

should we raise an exception here, or maybe just not continue and let python raise an exception on the next block where we try to use it as a dictionary?